packages/app/src/heartbeatService.ts

```diff
@@ -120,19 +127,80 @@ export class HeartbeatServiceImpl implements HeartbeatService {
     if (this._heartbeatsCache === null) {
       return '';
     }
+    // Extract as many heartbeats from the cache as will fit under the size limit.
+    const { heartbeatsToSend, unsentEntries } = extractHeartbeatsForHeader(
+      this._heartbeatsCache
+    );
     const headerString = base64Encode(
-      JSON.stringify({ version: 2, heartbeats: this._heartbeatsCache })
+      JSON.stringify({ version: 2, heartbeats: heartbeatsToSend })
     );
-    this._heartbeatsCache = null;
-    // Do not wait for this, to reduce latency.
-    void this._storage.deleteAll();
+    if (unsentEntries.length > 0) {
+      // Store any unsent entries if they exist.
+      this._heartbeatsCache = unsentEntries;
+      // This seems more likely than deleteAll (below) to lead to some odd state
+      // since the cache isn't empty and this will be called again on the next request,
+      // and is probably safest if we await it.
+      await this._storage.overwrite(this._heartbeatsCache);
+    } else {
+      this._heartbeatsCache = null;
+      // Do not wait for this, to reduce latency.
+      void this._storage.deleteAll();
+    }
     return headerString;
   }
 }
 
 function getUTCDateString(): string {
   const today = new Date();
-  return today.toISOString().substring(0,10);
+  return today.toISOString().substring(0, 10);
```

Comment:
(small nit) maybe add a comment saying what part of the date string is retained after the substring operation?

Reply:
Good idea, done.